### PR TITLE
#57: Task to verify the presence of headers

### PR DESCRIPTION
**Pull request by britter**
Fix for #57: Add task checkHeaders which will cause the build to fail in case there are files missing a header.

### CODE REVIEW

File: src/main/scala/de/heikoseeberger/sbtheader/HeaderPlugin.scala [hunk 1]
```diff
@@ -103,6 +106,13 @@ object HeaderPlugin extends AutoPlugin {
       ),
       headers.value,
       streams.value.log
+    ),
+    checkHeaders := checkHeadersTask(
+      FileFilter(Seq(excludes.value: _*)).filter(
+        (unmanagedSources in createHeaders).value.toList ++ (unmanagedResources in createHeaders).value.toList
+      ),
```
Comment (britter) [inline src/main/scala/de/heikoseeberger/sbtheader/HeaderPlugin.scala]: This block is redundant with the call to `createHeaders` above. Is is possible to assign the result to a local variable and reuse it for both calls? Or will this mess up with the task execution of sbt?

File: src/main/scala/de/heikoseeberger/sbtheader/HeaderPlugin.scala [hunk 2]
```diff
@@ -127,4 +137,23 @@ object HeaderPlugin extends AutoPlugin {
     log.debug(s"About to create/update header for $file")
     HeaderCreator(headerPattern, headerText, log, new FileInputStream(file)).getText.map(write)
   }
+
+  private def checkHeadersTask(files: Seq[File], headers: Map[String, (Regex, String)], log: Logger) = {
+    val filesWithoutHeader = files
+      .groupBy(_.extension)
+      .collect { case (Some(ext), groupedFiles) => headers.get(ext).map(_ -> groupedFiles) }
+      .flatten
+      .flatMap { case ((pattern, text), groupedFiles) => groupedFiles.flatMap(checkHeader(pattern, text, log)) }
```
Comment (britter) [inline src/main/scala/de/heikoseeberger/sbtheader/HeaderPlugin.scala]: Structurally redundant with the definition of `createHeaders`. I didn't find a way to better abstract the commons parts. Ideas?

File: src/main/scala/de/heikoseeberger/sbtheader/HeaderPlugin.scala [hunk 3]
```diff
@@ -127,4 +137,23 @@ object HeaderPlugin extends AutoPlugin {
...
+    if (filesWithoutHeader.nonEmpty) sys.error(
+      s"""|There are files without headers!
+          |  ${filesWithoutHeader.mkString(s"$newLine  ")}}
+          |""".stripMargin
+    )
+    filesWithoutHeader
+  }
+
+  private def checkHeader(headerPattern: Regex, headerText: String, log: Logger)(file: File) =
+    HeaderCreator(headerPattern, headerText, log, new FileInputStream(file)).getText.map(_ => file)
```
Comment (britter) [inline src/main/scala/de/heikoseeberger/sbtheader/HeaderPlugin.scala]: Kind of an abuse of the header creator...

File: src/main/scala/de/heikoseeberger/sbtheader/HeaderPlugin.scala (hunk 1 above)
Comment (hseeberger) [inline src/main/scala/de/heikoseeberger/sbtheader/HeaderPlugin.scala]: Needs to be duplicated due to `.value` macro magic.

File: src/main/scala/de/heikoseeberger/sbtheader/HeaderPlugin.scala (hunk 2 above)
Comment (hseeberger) [inline src/main/scala/de/heikoseeberger/sbtheader/HeaderPlugin.scala]: Why not put the common part into a method?

File: src/main/scala/de/heikoseeberger/sbtheader/HeaderPlugin.scala (hunk 1 above)
Comment (britter) [inline src/main/scala/de/heikoseeberger/sbtheader/HeaderPlugin.scala]: ok

File: src/main/scala/de/heikoseeberger/sbtheader/HeaderPlugin.scala (hunk 2 above)
Comment (britter) [inline src/main/scala/de/heikoseeberger/sbtheader/HeaderPlugin.scala]: done

**Comment by britter** [comment]
@hseeberger ready for merge from my PoV.

File: src/main/scala/de/heikoseeberger/sbtheader/HeaderPlugin.scala [hunk 4]
```diff
@@ -19,8 +19,10 @@ package de.heikoseeberger.sbtheader
 import java.io.File
 import java.nio.charset.StandardCharsets.UTF_8
 import java.nio.file.Files
+
 import sbt._
-import sbt.Keys._
+import sbt.Keys.{ unmanagedSources, _ }
```
Comment (hseeberger) [inline src/main/scala/de/heikoseeberger/sbtheader/HeaderPlugin.scala]: What? `_` should already bring all keys into scope.

File: src/main/scala/de/heikoseeberger/sbtheader/HeaderPlugin.scala [hunk 5]
```diff
@@ -127,4 +134,27 @@ object HeaderPlugin extends AutoPlugin {
...
+
+    if (filesWithoutHeader.nonEmpty) sys.error(
+      s"""|There are files without headers!
+          |  ${filesWithoutHeader.mkString(s"$newLine  ")}
+          |""".stripMargin
+    )
+    filesWithoutHeader
+  }
+
+  private def groupFilesByHeader(files: Seq[File], headers: Map[String, (Regex, String)]) = {
```
Comment (hseeberger) [inline src/main/scala/de/heikoseeberger/sbtheader/HeaderPlugin.scala]: Nice!

File: src/main/scala/de/heikoseeberger/sbtheader/HeaderPlugin.scala (hunk 4 above)
Comment (britter) [inline src/main/scala/de/heikoseeberger/sbtheader/HeaderPlugin.scala]: Sorry, looks like IntelliJ was to ambitious when organizing imports :-) I'll revert that.

**Pull request merged**